stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -265,16 +412,34 @@ inline Eigen::Matrix<var, -1, -1> cholesky_decompose(
       accum_i = accum;
     }
   } else {
-    cholesky_block* baseVari = new cholesky_block(A, L_A);
-    size_t pos = 0;
-    for (size_type j = 0; j < L.cols(); ++j) {
-      for (size_type i = j; i < L.cols(); ++i) {
-        L.coeffRef(i, j).vi_ = baseVari->variRefL_[pos++];
+#ifdef STAN_OPENCL
+    if (L_A.rows()
+        > opencl_context.tuning_opts().cholesky_size_worth_transfer) {
+      cholesky_gpu* baseVari = new cholesky_gpu(A, L_A);
+      size_t pos = 0;
+      for (size_type j = 0; j < L.cols(); ++j) {
+        for (size_type i = j; i < L.cols(); ++i) {
+          L.coeffRef(i, j).vi_ = baseVari->variRefL_[pos++];
+        }
+        for (size_type k = 0; k < j; ++k)
+          L.coeffRef(k, j).vi_ = dummy;
       }
-      for (size_type k = 0; k < j; ++k)
-        L.coeffRef(k, j).vi_ = dummy;
+    } else {
+#endif
+      cholesky_block* baseVari = new cholesky_block(A, L_A);
+      size_t pos = 0;
+      for (size_type j = 0; j < L.cols(); ++j) {
+        for (size_type i = j; i < L.cols(); ++i) {
+          L.coeffRef(i, j).vi_ = baseVari->variRefL_[pos++];
+        }
+        for (size_type k = 0; k < j; ++k)
+          L.coeffRef(k, j).vi_ = dummy;
+      }
```

Comment:
Related to refactoring, if these all called one class and we had the chain method use the different algorithms it would clean the code up there a lot. Again not sure if that should be done in this PR or a separate one